### DISCUSSION TITLE
Past mode on heroku

### DISCUSSION THREAD

**Discussion by samadii** [aahnik/tgcf#125] [General]
There are some bots that can clone chats by deploying to heroku, so why the tgcf past mode can't work perfectly on heroku ?

**Comment by tissole** [comment]
The script keeps track of files forwarded in a config file, but Heroku dumps everything written there when the dyno restarts (after 24h or when stopped by the user). This is the so-called "ephemeral file system", you cannot keep any files modified after the dyno starts.

I heard that are some clever ways to bypass this by using a database, but I don't know if the scripts can use that.

**Comment by aahnik** [comment]
> The script keeps track of files forwarded in a config file

Specifically speaking the offset is updated when you run `tgcf` in past mode. 

and hi @Soheib1378, you can definitely run tgcf in a past mode in Heroku, but the offset will not be updated in your `TGCF_CONFIG` environment variable.

It is great for one-time running if you don't care about offset.

GitHub actions can be used to run `tgcf` in the past mode for free. I will write the tutorial soon!

**Reply by ankit-sinha-18** [reply]
`tgcf` can be run on heroku in `past` mode, but when whole transfer would complete,then it would crash and send the same file again to those Channel which are in config file,
It means files will be duplicated if you wouldn't stop when the first time whole transfer will complete

**Reply by tissole** [reply]
Do you mean that when the dyno restart after 24h the script will be automatically restarted too? I was under the impression that you have to start the script manually to work after 24h.

**Reply by ankit-sinha-18** [reply]
Yep , script will auto restart if dynos restart in 24 hours but ,
There is a problem,when it would restart it will forgot the updated `tgcf.config.yml` and Since Offset id is old message id so ,it will start forwarding the same old message as configured in `TGCF_CONFIG` variable.

**Reply by tissole** [reply]
I understand. If offset is set to zero initially, the whole channel will be cloned again, and if the offset is set to 2000 al the msgs from that id up will be forwarded again. But if @aahnik will update the code to prevent forwarding duplicates, Heroku will be defeated :)

**Comment by tissole** [comment]
@Soheib1378 After 24h Heroku resets the offset value so you will not know the last message forwarded. But there is a way, go to the destination channel and look at the last message, and then look for that message in the source channel. Copy his id from the source and put it in the offset in your config. The next day the forwarding will start with the next message.

**Reply by samadii** [reply]
Thanks bro, I will try.

**Reply by aahnik** [reply]
hi @tissole, how will you copy the id of the message?

**Reply by ankit-sinha-18** [reply]
@aahnik if you will use the unofficial app of Telegram like plus messenger,then you can check the message id of the message by tapping on the message and check message details, 
And another way is ,tap on any message in any chat and you will see ,copy link option
In that link ,the last combination of digit will show the message id of that chat.

Link will be like for private Channel:-
https://t.me/c/7388389273/734

In this link 734 is the message id of Channel whose id is -1007388389273

**Reply by tissole** [reply]
For desktop, I use [Kotatogram](https://github.com/kotatogram/kotatogram-desktop). I can copy the message ID, but basically is the same as the link to the post as @ankit-sinha-18 explained. This client shows by default the channel ID too when you click on the channel name. Still has some bugs but is decent.

**Comment by u-8d** [comment]
What if I stopped the bot dyno after cloning.

**Reply by aahnik** [reply]
you can stop that. after the script finishes, Heroku will restart and continue forwarding from the beginning.

it is ideal to use past mode from the Heroku console. or from your computer.

will update the wiki on this!

**Reply by ankit-sinha-18** [reply]
But it is not possible to keep pc turn on for long period if using terminal in Heroku.
It will stop forwarding as soon as the window will closed

**Reply by aahnik** [reply]
> It will stop forwarding as soon as the window will closed

I thought processes keep running in background.

don't, worry, the github action is coming soon. that's the easiest and best way to use in past mode. and completely free forever. offset will be updated, and you can run one time, or periodically (like everyday). 

the script can run continuously for as long as 72 hours.

**Reply by tissole** [reply]
Heroku has 3 types of dynos:
- one-off dynos running from your console, you close console dyno gets offline
- web and worker dynos that run even you close your PC. Web dynos are for lite jobs, worker dynos are for heavy background jobs.
